botwo/botwo.py

```diff
@@ -137,52 +150,51 @@ def _api_call(*args, **kwargs):
+
+        _paginator_api_call.__name__ = str(operation_name)
+        return _paginator_api_call
+
+    def _create_can_paginate_method(self, operation_name):
+        def _can_paginate_api_call(*args, **kwargs):
+            return getattr(self.default, operation_name)(**kwargs)
+
+        _can_paginate_api_call.__name__ = str(operation_name)
+        return _can_paginate_api_call
 
-        def _delete_objects_api_call(*args, **kwargs):
+    def _create_delete_objects_method(self, operation_name):
+        def _delete_objects_api_call(_, *args, **kwargs):
             if args:
                 raise TypeError("%s() only accepts keyword arguments." % operation_name)
-            # TODO
-            return None
-
-        if operation_name == "get_paginator":
-            _paginator_api_call.__name__ = str(operation_name)
-            return _paginator_api_call
-        if operation_name == "delete_objects":
-            _delete_objects_api_call.__name__ = str(operation_name)
-            return _delete_objects_api_call
-
-    def _route_params(self, path, api_params, is_copy):
-        client_to_call = self.default_client
-        for map in self.mapping:
-            if fnmatch.fnmatch(path, map.get("bucket_name")):
-                if map.get("client"):
-                    if map.get("client") == self.client2_name:
-                        client_to_call = self.client2
-                    elif map.get("client") == self.client1_name:
-                        client_to_call = self.client1
-                    else:
-                        raise ValueError("Invalid client name " + map.get("client"))
-                else:
-                    raise ValueError("client isn't provided")
-
-                if not is_copy:
-                    if map.get("mapped_bucket_name"):
-                        api_params["Bucket"] = map.get("mapped_bucket_name")
-                    if map.get("key_prefix"):
-                        api_params["Key"] = map.get("key_prefix") + api_params["Key"]
-                else:
-                    source_path = api_params["CopySource"].split("/")
-                    source_bucket = source_path[0]
-                    source_key = "/".join(source_path[1::])
-                    if map.get("mapped_bucket_name"):
-                        source_bucket = map.get("mapped_bucket_name")
-                    if map.get("key_prefix"):
-                        source_key = map.get("key_prefix") + api_params["Key"]
-                    api_params["CopySource"] = source_bucket + "/" + source_key
-                break
-        return client_to_call, api_params
+            if kwargs.get("Delete"):  # delete objects operation
+                for i, obj in enumerate(kwargs["Delete"]["Objects"]):
+                    params = {"Bucket": kwargs.get("Bucket"), "Key": obj["Key"]}
+                    res = self._route_params(params)
+                    client_to_call = res[0]
+                    bucket = res[1]["Bucket"]
```

Comment:
Can we do this instead?

```suggestion
                    client_to_call, b = self._route_params(params)
                    bucket = b["Bucket"]
```
(also in all calls to `_route_params`).

Also I think other places destructure like this -- and it is quite hard for me to read.

Reply:
Done